File: src/main/java/org/example/eventhub/specification/EventSpecification.java
package org.example.eventhub.specification;

import jakarta.persistence.criteria.Predicate;
import org.example.eventhub.dto.event.EventFilter;
import org.example.eventhub.entity.Event;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Component
public class EventSpecification {

    public Specification<Event> withFilter(EventFilter filter) {
        return (root, query, cb) -> {
           if (filter == null) {
               return cb.conjunction();
           }

            List<Predicate> predicates = new ArrayList<>();

            if (filter.title() != null && !filter.title().isBlank()) {
                String lowerTitle = filter.title().toLowerCase();
                predicates.add( cb.like(
                        cb.lower(root.get("title")),
                        "%" + lowerTitle + "%"
                ));
            }

            if (filter.city() != null && !filter.city().isBlank()) {
                String lowerCity = filter.city().toLowerCase();
                predicates.add( cb.like(
                        cb.lower(root.get("location").get("city")),
                        "%" + lowerCity + "%"
                ));
            }

            if (filter.minCapacity() != null) {
                predicates.add(cb.greaterThanOrEqualTo(root.get("capacity"), filter.minCapacity()));
            }
            if (filter.maxCapacity() != null) {
                predicates.add(cb.lessThanOrEqualTo(root.get("capacity"), filter.maxCapacity()));
            }

            if (filter.minPrice() != null) {
                predicates.add(cb.greaterThanOrEqualTo(root.get("price"), filter.minPrice()));
            }
            if (filter.maxPrice() != null) {
                predicates.add(cb.lessThanOrEqualTo(root.get("price"), filter.maxPrice()));
            }

            if (filter.eventStatus() != null) {
                predicates.add(cb.equal(root.get("eventStatus"), filter.eventStatus()));
            }

            if (filter.fromDateTime() != null) {
                predicates.add(cb.greaterThanOrEqualTo(
                        root.get("dateTime"),
                        filter.fromDateTime())
                );
            }

            if (filter.toDateTime() != null) {
                predicates.add(cb.lessThanOrEqualTo(
                        root.get("dateTime"),
                        filter.toDateTime())
                );
            }

            if (filter.upcoming() != null && filter.upcoming()) {
                predicates.add(cb.greaterThanOrEqualTo(
                        root.get("dateTime"),
                        LocalDateTime.now())
                );
            }

            return cb.and(predicates.toArray(new Predicate[0]));

        };
    }
}

File: src/main/kotlin/org/example/eventhub/specification/EventSpecification.kt
package org.example.eventhub.specification

import jakarta.persistence.criteria.Predicate
import org.example.eventhub.dto.event.EventFilter
import org.example.eventhub.entity.Event
import org.example.eventhub.entity.Location
import org.example.eventhub.enums.EventStatus
import org.springframework.data.jpa.domain.Specification
import org.springframework.stereotype.Component
import java.time.LocalDateTime

@Component
class EventSpecification {

    fun withFilter(filter: EventFilter?): Specification<Event> =
        Specification { root, _, cb ->

            if (filter == null) {
                return@Specification cb.conjunction()
            }

            val predicates = mutableListOf<Predicate>()

            filter.title
                ?.takeIf { it.isNotBlank() }
                ?.let {
                    predicates += cb.like(
                        cb.lower(root.get("title")),
                        "%${it.lowercase()}%"
                    )
                }

            filter.city
                ?.takeIf { it.isNotBlank() }
                ?.let {
                    predicates += cb.like(
                        cb.lower(root.get<Location>("location").get("city")),
                        "%${it.lowercase()}%"
                    )
                }

            filter.minCapacity?.let {
                predicates += cb.greaterThanOrEqualTo(root.get("capacity"), it)
            }

            filter.maxCapacity?.let {
                predicates += cb.lessThanOrEqualTo(root.get("capacity"), it)
            }

            filter.minPrice?.let {
                predicates += cb.greaterThanOrEqualTo(root.get("price"), it)
            }

            filter.maxPrice?.let {
                predicates += cb.lessThanOrEqualTo(root.get("price"), it)
            }

            filter.eventStatus?.let {
                predicates += cb.equal(root.get<EventStatus>("eventStatus"), it)
            }

            filter.fromDateTime?.let {
                predicates += cb.greaterThanOrEqualTo(root.get("dateTime"), it)
            }

            filter.toDateTime?.let {
                predicates += cb.lessThanOrEqualTo(root.get("dateTime"), it)
            }

            if (filter.upcoming == true) {
                predicates += cb.greaterThanOrEqualTo(
                    root.get("dateTime"),
                    LocalDateTime.now()
                )
            }

            cb.and(*predicates.toTypedArray())
        }
}
